### PR TITLE
Lazy/Eager evaluation of YAML tag nodes

### LINKED ISSUE #99: YAML tags do not work with nested mapping or sequence nodes
Currently the COBalD YAML tags do not work with nested mapping or sequence nodes like:

```yaml
!Foo
Bar:
  - Bla: Blubb
```
which constructs a python object `Foo(Bar=[])` instead of `Foo(Bar=[{"Bla": "Blubb"}])`. 

The later can be archieved by adding the `deep=True` argument to `construct_mapping` in https://github.com/MatterMiners/cobald/blob/094653385f942fa9f912c2602b8d1ff4ba7419e9/src/cobald/daemon/config/yaml.py#L50 and `construc_sequence` in https://github.com/MatterMiners/cobald/blob/094653385f942fa9f912c2602b8d1ff4ba7419e9/src/cobald/daemon/config/yaml.py#L55.

Could this be added to COBalD YAML tags, please?

### PR DESCRIPTION
**Pull request by maxfischer2781**
This PR addresses the evaluation strategy of YAML tags using nested nodes.

- added tests for lazy/eager evaluation
- ``yaml_constructor`` takes an optional argument for eager evaluation
- added a decorator for settings of YAML tags

Closes #99. FYI @giffels 

------------

Announcing a plugin still stays the same, e.g.
```
entry_points={
        "cobald.config.yaml_constructors": [
            "TardisRestApi = tardis.rest.service:RestService",
        ],
```
but the actual code object can be decorated to apply settings:
```
from cobald.daemon.plugins import yaml_tag
....

@service(flavour=asyncio)
@yaml_tag(eager=True)  # <<< magic sauce
class RestService(object):
```

### CODE REVIEW

File: cobald_tests/daemon/core/test_config.py [hunk 1]
```diff
@@ -13,6 +15,60 @@
 COBalDLoader.add_constructor(tag="!MockPool", constructor=yaml_constructor(MockPool))
 
 
+class TagTracker:
+    """Helper to track the invocation of YAML !Tags"""
```
Comment (eileen-kuehn) [inline cobald_tests/daemon/core/test_config.py:19]: I don't get what this class is actually supposed to do from within the class definition. Could you maybe explain and extend the docstring to actually document how this is supposed to work/being used?

File: cobald_tests/daemon/core/test_config.py [hunk 2]
```diff
@@ -13,6 +15,60 @@
...
+
+
+@yaml_constructor(eager=True)
+def full_deco_constructor(*args, **kwargs):
+    return copy.deepcopy(args), copy.deepcopy(kwargs)
+
+
+@yaml_constructor
+def bare_deco_constructor(*args, **kwargs):
+    return copy.deepcopy(args), copy.deepcopy(kwargs)
```
Comment (eileen-kuehn) [inline cobald_tests/daemon/core/test_config.py:55]: How does the naming tell me what is done? The only difference I can spot is that one is used lazy and the other eager. Why is one then called `full` and the other `bare`?

File: cobald_tests/daemon/core/test_config.py [hunk 3]
```diff
@@ -95,10 +151,141 @@ def test_load_mixed_creation(self):
...
+                            - leaf: "leaf level value"
+                    """
+                )
+            with load(config.name) as config:
+                tagged = get_config_section(config, "__config_test__")["tagged"]
+                assert isinstance(tagged, TagTracker)
+                assert tagged.orig_kwargs["top"] == "top level value"
+                assert isinstance(tagged.orig_kwargs["nested"], list)
+                assert len(tagged.orig_kwargs["nested"]) > 0
+                assert tagged.orig_kwargs["nested"] == [{"leaf": "leaf level value"}]
```
Comment (eileen-kuehn) [inline cobald_tests/daemon/core/test_config.py:236]: Do you have to use `orig_kwargs` here or would `kwargs` also work?

**Review by eileen-kuehn — CHANGES_REQUESTED** [review]
Some parts in the unit tests are hard to understand and might improve readability by adding some comments. 
Otherwise I think it is good to go 👍 Great stuff you spotted this...

File: cobald_tests/daemon/core/test_config.py (hunk 3 above)
Comment (maxfischer2781) [reply cobald_tests/daemon/core/test_config.py:231]: I need ``orig_kwargs`` to verify that all the data was already there during YAML parsing.

File: cobald_tests/daemon/core/test_config.py [hunk 4]
```diff
@@ -13,6 +14,57 @@
...
+class TagTracker:
+    """Helper to track the arguments supplied to YAML !Tags"""
+
+    def __init__(self, *args, **kwargs):
+        # the state of arguments *during* YAML evaluation
+        self.orig_args = copy.deepcopy(args)
+        self.orig_kwargs = copy.deepcopy(kwargs)
+        # the state of arguments *after* YAML evaluation
+        self.final_args = args
+        self.final_kwargs = kwargs
```
Comment (giffels) [inline cobald_tests/daemon/core/test_config.py:29]: This a bit weird on the first glance, but I have no really idea to make it clearer.

**Pull request merged**